binary_search_tree/tree.py

```diff
@@ -14,46 +17,187 @@ class Tree:
+        
+    def inorder_helper(self, cur, arr):
+        if cur == None:
+            return arr
+        else:
+            left_arr = self.inorder_helper(cur.left, arr)
+            left_arr.append({ 
+                            "key" : cur.key, 
+                            "value" : cur.value
+                            })
+            right_arr = self.inorder_helper(cur.right, left_arr)
+            return right_arr
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def preorder(self):
-        pass
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
+    def preorder(self): # root, left, right
+        arr = []
+        if self.root == None:
+            return arr
+        else:
+            return self.preorder_helper(self.root, arr)
+    
+    
+    def preorder_helper(self, cur, arr):
+        if cur == None:
+            return arr
+        else:
+            cur_dict = { 
+                        "key" : cur.key, 
+                        "value" : cur.value
+                        }
+            # root
+            arr.append(cur_dict)
+            # left
+            left_arr = self.preorder_helper(cur.left, arr)
+            
+            # right
+            right_arr = self.preorder_helper(cur.right, left_arr)
+            return right_arr
 
-    # Time Complexity: 
-    # Space Complexity:     
-    def postorder(self):
-        pass
+    # Time Complexity: O(n)
+    # Space Complexity:   O(n)  
+    def postorder(self): # left, right, root
+        arr = []
+        if self.root == None:
+            return arr
+        else:
+            return self.postorder_helper(self.root, arr)
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper(self, cur, arr):
+        if cur == None:
+            return arr
+        else:
+            left_arr = self.postorder_helper(cur.left, arr)
+            right_arr = self.postorder_helper(cur.right, left_arr)
+            right_arr.append({ 
+                        "key" : cur.key, 
+                        "value" : cur.value
+                        })
+            return right_arr
+
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def height(self):
-        pass
+        if self.root == None:
+            return 0 
+        else:
+            return self.height_helper(self.root)
+    
+    def height_helper(self, cur):
+        max_height = 0
+        if cur == None: # 5 3 None
+            return max_height 
+        
+        else:
+            left_height = self.height_helper(cur.left) 
+            right_height = self.height_helper(cur.right) 
+            max_height = max(left_height, right_height)
+            return max_height + 1
 
 
 #   # Optional Method
-#   # Time Complexity: 
-#   # Space Complexity: 
+#   # Time Complexity: O(n)
+#   # Space Complexity: O(n)
+# bfs - This method returns an array with the tree elements in a level-by-level order 
     def bfs(self):
```

Comment:
👍 Just FYI because with lists pop(0) is an O(n) operation, this has time complexity O(n^2).